File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Common/PumpStatus.cs
using System.Text.Json.Serialization;

namespace FccDesktopAgent.Core.Adapter.Common;

/// <summary>
/// Pump nozzle status snapshot from the FCC or synthesized by the Edge Agent.
/// </summary>
public sealed record PumpStatus
{
    [JsonPropertyName("schemaVersion")]
    public string SchemaVersion { get; init; } = "1.0";

    [JsonPropertyName("siteCode")]
    public string SiteCode { get; init; } = string.Empty;

    [JsonPropertyName("pumpNumber")]
    public int PumpNumber { get; init; }

    [JsonPropertyName("nozzleNumber")]
    public int NozzleNumber { get; init; }

    [JsonPropertyName("state")]
    public PumpState State { get; init; }

    [JsonPropertyName("currencyCode")]
    public string CurrencyCode { get; init; } = string.Empty;

    [JsonPropertyName("statusSequence")]
    public long StatusSequence { get; init; }

    [JsonPropertyName("observedAtUtc")]
    public DateTimeOffset ObservedAtUtc { get; init; }

    [JsonPropertyName("source")]
    public PumpStatusSource Source { get; init; }

    [JsonPropertyName("productCode")]
    public string? ProductCode { get; init; }

    [JsonPropertyName("productName")]
    public string? ProductName { get; init; }

    /// <summary>Decimal string representation of current volume in litres.</summary>
    [JsonPropertyName("currentVolumeLitres")]
    public string? CurrentVolumeLitres { get; init; }

    /// <summary>Decimal string representation of current dispensed amount.</summary>
    [JsonPropertyName("currentAmount")]
    public string? CurrentAmount { get; init; }

    /// <summary>Decimal string representation of unit price.</summary>
    [JsonPropertyName("unitPrice")]
    public string? UnitPrice { get; init; }

    [JsonPropertyName("fccStatusCode")]
    public string? FccStatusCode { get; init; }

    [JsonPropertyName("lastChangedAtUtc")]
    public DateTimeOffset? LastChangedAtUtc { get; init; }

    /// <summary>
    /// Extended supplemental status from FpStatus_3 (16 parameter IDs).
    /// Null when the FCC does not include supplemental data in the response.
    /// </summary>
    [JsonPropertyName("supplemental")]
    public PumpStatusSupplemental? Supplemental { get; init; }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/common/PumpStatus.kt
package com.fccmiddleware.edge.adapter.common

import kotlinx.serialization.Serializable

/**
 * Real-time pump state snapshot from the FCC.
 *
 * Matches pump-status.schema.json v1.0 field-for-field.
 * In-flight decimal amounts (volume, price) are represented as formatted strings
 * per the schema — they are display values, not arithmetic operands.
 * All monetary arithmetic uses Long minor units elsewhere.
 */
@Serializable
data class PumpStatus(
    /** Schema version. Always "1.0". */
    val schemaVersion: String = "1.0",

    /** Site identifier. */
    val siteCode: String,

    /** Physical pump number. */
    val pumpNumber: Int,

    /** Physical nozzle number. */
    val nozzleNumber: Int,

    /** Current operational state of the pump-nozzle pair. */
    val state: PumpState,

    /** ISO 4217 currency code. */
    val currencyCode: String,

    /** Monotonically increasing sequence number for ordering concurrent status updates. */
    val statusSequence: Int,

    /** When this status was observed. UTC ISO 8601 with trailing Z. */
    val observedAtUtc: String,

    /** Whether this reading came directly from the FCC or was synthesised by the Edge Agent. */
    val source: PumpStatusSource,

    /** Canonical product code currently on the nozzle. Null when idle or unknown. */
    val productCode: String? = null,

    /** Human-readable product name. Null when idle or unknown. */
    val productName: String? = null,

    /**
     * In-progress dispensed volume in litres as a decimal string (e.g. "12.345").
     * Null when not dispensing. Pattern: ^(0|[1-9][0-9]*)(\.[0-9]{1,3})?$
     */
    val currentVolumeLitres: String? = null,

    /**
     * In-progress dispensed amount as a decimal string (e.g. "152.50").
     * Null when not dispensing. Pattern: ^(0|[1-9][0-9]*)(\.[0-9]{1,2})?$
     */
    val currentAmount: String? = null,

    /**
     * Unit price per litre as a decimal string (e.g. "12.3400").
     * Null when not available. Pattern: ^(0|[1-9][0-9]*)(\.[0-9]{1,4})?$
     */
    val unitPrice: String? = null,

    /** Vendor-native status code for diagnostics. Null if not exposed by FCC. */
    val fccStatusCode: String? = null,

    /** When the pump state last changed. Null if vendor does not report. UTC ISO 8601 with trailing Z. */
    val lastChangedAtUtc: String? = null,

    /** Extended supplemental status from FpStatus_3. Null when not included. */
    val supplemental: PumpStatusSupplemental? = null,
)
